text/3668-async-closure.md

```diff
@@ -0,0 +1,713 @@
+}
+```
+
+# Detailed Explanation
+
+### `AsyncFn*`
+
+This RFC introduces a family of `AsyncFn` traits. These traits are intended to remain unstable to name or implement, just like the `Fn` traits. Nonetheless, we'll describe the details of these traits so as to explain the user-facing features enabled by them.
+
+The definition of the traits is (modulo `rustc_` attributes, and the `"rust-call"` ABI):
+
+
+> [!NOTE]
+> We omit some details about the `"rust-call"` calling convention and the fact that the `Args` parameter is enforced to be a tuple.
+
+```rust
+/// An async-aware version of the [`FnOnce`](crate::ops::FnOnce) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnOnce::async_call_once`].
+    type CallOnceFuture: Future<Output = Self::Output>;
+
+    /// Output type of the called closure's future.
+    type Output;
+
+    /// Call the [`AsyncFnOnce`], returning a future which may move out of the called closure.
+    fn async_call_once(self, args: Args) -> Self::CallOnceFuture;
+}
+
+/// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnMut<Args>: AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnMut::async_call_mut`] and [`AsyncFn::async_call`].
+    type CallRefFuture<'a>: Future<Output = Self::Output>
+    where
+        Self: 'a;
+
+    /// Call the [`AsyncFnMut`], returning a future which may borrow from the called closure.
+    fn async_call_mut(&mut self, args: Args) -> Self::CallRefFuture<'_>;
+}
+
+/// An async-aware version of the [`Fn`](crate::ops::Fn) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFn<Args>: AsyncFnMut<Args> {
+    /// Call the [`AsyncFn`], returning a future which may borrow from the called closure.
+    fn async_call(&self, args: Args) -> Self::CallRefFuture<'_>;
+}
+```
+
+### Associated types of `AsyncFn*` traits are not nameable
+
+Unlike what is true today with the current `Fn*` traits, this RFC reserves as an implementation detail the associated types of the `AsyncFn*` traits, and these will not be nameable as part of the stable interface specified by this RFC.
+
+That is, using the existing `FnOnce` trait, we can write this today on stable Rust:
+
+```rust
+fn foo<F, T>()
+where
+    F: FnOnce() -> T,
+    F::Output: Send, //~ OK
+{
+}
+```
+
+(We decided to allow this in [#34365](https://github.com/rust-lang/rust/pull/34365).)
+
+However, this RFC reserves as an implementation detail the associated types of the traits specified above, so this does not work:
+
+```rust
+fn foo<F, T>()
+where
+    F: async FnOnce() -> T,
+    F::Output: Send,
+    //~^ ERROR use of unstable library feature
+    F::CallOnceFuture: Send,
+    //~^ ERROR use of unstable library feature
+{
+}
+```
+
+### `async` bound modifier on `Fn()` trait bounds
```

Comment:
@joshtriplett responding inline where the topic of your concern is

While I disagree with the result of your analysis and your rationale, it's a valid concern.

It is nice to have syntactical consistency between `async fn` items and `async Fn` traits, but I also see your point on this RFC handling two topics at the same time, even though the core feature is only one of them

One thing though:

> we manage to completely replace the traits with `LendingFn*` instead, we can make the `AsyncFn*` syntax be an alias for that and continue working.

That's going to require quite some hacks (or keeping the traits around just for the syntax to work, or potentially some heroics with trait aliases) to support, as currently the `Fn(Foo) -> Bar` syntax works for all traits at the syntax *and name resolution* level, and is only rejected after we resolve the trait to one that does not support this syntax.

The new syntax on the other hand gives us a lot of freedom on the desugaring phase, and thus avoids having to think about lending iterators or backwards compatibly changing Fn traits at this time.

Could your concern be resolved by making it is only part of the parenthesized Fn* trait sugar, so it doesn't prescribe any behaviour for `async Trait`, and doesn't even parse that syntax anymore. Right now

```rust
#![feature(async_closure)]
#[cfg(FALSE)]
fn foo<T: async Iterator>() {}
```

parses (see [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=56f0a7220f225cdbe56747b77ea4bf8f)), but we could make it so only

```rust
#![feature(async_closure)]
#[cfg(FALSE)]
fn foo<T: async Iterator()>() {}
```

parses (just like `T: Iterator()` parses on stable)

Reply:
@oli-obk:
>> we can make the `AsyncFn*` syntax be an alias for that and continue working.
>
> That's going to require quite some hacks (or keeping the traits around just for the syntax to work, or potentially some heroics with trait aliases) to support, as currently the `Fn(Foo) -> Bar` syntax works for all traits at the syntax and name resolution level, and is only rejected after we resolve the trait to one that does not support this syntax.

I don't think it'd be *that* awful if we ended up, in the *very worst* case, keeping the `AsyncFn*` traits around as separate traits, and separately providing `LendingFn`. I'm hoping we can do better than that via some kind of alias or alias-like hack, but the fact that we could always fall back to a separate trait means this wouldn't be a fatal problem.

I did expect that an alias mechanism would require keeping at least the *name* around in the standard library in some form, whether an alias or some `rustc_old_async_fn_trait`-annotated magic. Would that suffice, such that we resolve the name and notice that it's an `AsyncFn` trait marked with some `rustc_old_async_fn_trait` label and map the syntax accordingly? Is that the part that would be a hack, or would you expect the required hackery in the compiler to be substantially worse than that?

In any case, I agree that that would not be a wonderful transition to go through. The details of that transition would presumably depend on what `LendingFn*` ends up looking like. I've been assuming that we're going to end up needing `LendingFn` anyway in order to ship `gen ||`, rather than having to ship a `GenFn` trait.

(As an aside, I'm really looking forward to seeing what general "lending closures" look like!)

> Could your concern be resolved by making it is only part of the parenthesized `Fn*` trait sugar, so it doesn't prescribe any behaviour for `async Trait`

No, that wouldn't help here. From a user-visible perspective I don't think that makes any difference; they'll still see `async Fn(...)`, which still looks like a trait modifier on `Fn` rather than `AsyncFn`, which still raises the concerns I described.

Reply:
> Would that suffice, such that we resolve the name and notice that it's an `AsyncFn` trait marked with some `rustc_old_async_fn_trait` label and map the syntax accordingly? Is that the part that would be a hack, or would you expect the required hackery in the compiler to be substantially worse than that?

It's either going to be a hack in name resolution that I don't think is going to be accepted, or a lang item remapping hack at ast lowering time, which, while fairly contained, would still be an awful wart I would like to avoid



> which still looks like a trait modifier on `Fn` rather than `AsyncFn`

Right. I think this is the core point of our disagreement. I think that's better than a separate trait exposed to end users. I think everything from diagnostics to rustdoc can be made significantly nicer than two separate, almost-the-same traits. We could absolutely make the backing traits internal with no way to access them even in libcore, effectively making them a compiler internal implementation detail that we can replace to make the feature more powerful (hypothetical keyword generics come to mind, or more realistically special cased `dyn async Fn()` support ahead of general dyn support for arbitrary impl-trait-in-methods, or likely just something unanticipated we'll notice with wide spread usage and want to fix).

Reply:
It's not impossible that in the future we'll realize we'll actually want individual traits for some specific use cases. We'll have the option to move to exposing these traits (or readding if we removed them). I think even then the common case of these being the `async` version of `Fn*` traits will be better syntactically, in diagnostics and rustdoc.

Reply:
> or a lang item remapping hack at ast lowering time, which, while fairly contained, would still be an awful wart I would like to avoid

Completely agreed. I would like to avoid that as well. I wanted to know if we were talking something *worse* than that or "just" an unpleasant wart.

(I think the best-case scenario would be if we end up shipping LendingFn before we need to stabilize this. But I don't necessarily want to block stabilizing this feature on that, unless `LendingFn` is on a shorter-term horizon than I've gotten the impression it is.)

> > which still looks like a trait modifier on `Fn` rather than `AsyncFn`

> Right. I think this is the core point of our disagreement.

Yeah, I agree.

I think *if* we had some consensus that we definitely want to have `async Trait` in general, and roughly what it would mean, then I'd be on board with having a preliminary version of it here. But we don't have that consensus.

Reply:
From my perspective, providing a convenient shorthand for a special case of a complicated trait bound is what trait aliases are _for_. So if we need to move `AsyncFn` to be a trait alias in the future, I don’t see how that’s a problem, it’s using an existing language feature for its intended purpose! Which seems preferable to adding a new, redundant feature like `async` modifiers.

Reply:
> I think _if_ we had some consensus that we definitely want to have `async Trait` in general, and roughly what it would mean, then I'd be on board with having a preliminary version of it here. But we don't have that consensus

The parenthesized trait bound syntax for the Fn traits has always been magical and meant to look like function items and pointers. Moving to a new trait diverges from that. It's more consistent imo to extend the existing magic Fn bounds than to add a new trait. I also don't think it's suprising to anyone that the `Fn` trait bounds are different from regular trait bounds, and making them more different (but more similar to the corresponding fn decls), is in fact the consistent thing to do

Reply:
> (I think the best-case scenario would be if we end up shipping LendingFn before we need to stabilize this. But I don't necessarily want to block stabilizing this feature on that, unless `LendingFn` is on a shorter-term horizon than I've gotten the impression it is.)

Are you suggesting if we had `LendingFn` right now we should use them instead of `AsyncFn`? I think that would be confusing to users and very verbose. Like forcing everyone to use `impl Future` instead of `async fn`

Reply:
> if we need to move `AsyncFn` to be a trait alias in the future, I don’t see how that’s a problem

I'm saying that at this stage no one can actually promise that will be possible. It's not clear that lending will be compatible enough for the type system to handle that. A sugar on the other hand has so much more freedom that we can be compatible with any future thing while providing excellent diagnostics. Diagnostics and Rustdoc will be subpar or fragile if we go with just the traits and trait aliases even if the trait aliases could work out (not holding my breath considering how long we've had trouble with them)

Reply:
To put the consistency argument another way, `AsyncFn(...) -> T` would make the `Fn` trait sugar behave differently depending on the trait name. Even in a world where we wind up without general `async Trait` and with stable `AsyncFn` and/or `LendingFn`, that would *still* be a wart.

The alternative warts we're weighing this against are either a) that we have *only* `async Fn` but no other `async Trait`, or b) that we have `async Fn` and `async Trait` but they are somehow subtly different. Both of these are risks of the `async Fn` sugar, but `AsyncFn(...) -> T` would have the above wart *today.*

Reply:
> would make the `Fn` trait sugar behave differently depending on the trait name.

Not really. For every item (`trait` or `trait` alias) supporting the `->` sugar, what follows the `->` corresponds to some associated type; which associated type this is is a component of the item definition.

Reply:
> > (I think the best-case scenario would be if we end up shipping LendingFn before we need to stabilize this. But I don't necessarily want to block stabilizing this feature on that, unless `LendingFn` is on a shorter-term horizon than I've gotten the impression it is.)
> 
> Are you suggesting if we had `LendingFn` right now we should use them instead of `AsyncFn`? I think that would be confusing to users and very verbose. Like forcing everyone to use `impl Future` instead of `async fn`

No, not necessarily; I'm saying that if `LendingFn*` were done and its exact interface was something we were ready to stabilize, we'd evaluate the ideal syntactic sugar in terms of *that* rather than stopgap `AsyncFn*` traits. What precisely that sugar should look like is another question.

At a minimum, it'd be helpful to see what precisely the full LendingFn bounds for a function accepting a closure would be.

Reply:
> Not really. For every item (trait or trait alias) supporting the -> sugar, what follows the -> corresponds to some associated type; which associated type this is is a component of the item’s definition.

This is exactly what I'm saying it would *become*. Today it is always the `FnOnce::Output` associated type; with `AsyncFn(...) -> T` it would be different depending on the trait. With `async Fn(...) -> T` the choice of `FnOnce::Output` vs <unspecified async assoc type> would depend only on the sugar syntax.

Reply:
> At a minimum, it'd be helpful to see what precisely the full LendingFn bounds for a function accepting a closure would be.

I assume you mean "accepting an **async** closure". In that case, I did share that in the RFC: https://github.com/compiler-errors/rust-rfcs/blob/async-closure-redux/text/3668-async-closure.md#why-do-we-need-a-new-set-of-asyncfn-traits

Specifically, in its current design, a `LendingFn`-based `async Fn()` bound would need to desugar into a two associated type bounds:

```
where F: async Fn() -> i32

// is

where F: for<'s> LendingFn<LendingOutput<'s>: Future<Output = i32>> + FnOnce<Output: Future<Output = i32>>
```

Reply:
> I'm saying that at this stage no one can actually promise that will be possible. It's not clear that lending will be compatible enough for the type system to handle that. A sugar on the other hand has so much more freedom that we can be compatible with any future thing while providing excellent diagnostics.

Even in the worst case, where none of the trait aliases infrastructure can be used, and `AsyncFn` ends up desugaring to some completely custom special‐case thing—that will be no different, implementation‐wise, from `async Fn` desugaring to that same completely custom special‐case thing, no? So our freedom is not restricted in any way.

Reply:
There is a major difference in hacks needed in the compiler.

Also just supporting the base `AsyncFn()` case requires hacks due to the `-> Type` specifying the `Future::Output` type instead of the type of the future (which makes sense, but is such a divergence from the trivial desugaring that is `Fn()`, it will be painful to support).

I completely forgot about that part, thx @rpjohnst for bringing it up again. Basically I think explaining how `AsyncFn()` desugars differently from `Fn()` will be significantly more confusing than explaining the `async Fn()` syntax, which is just "take this other branch in the compiler". Explicit syntax helps guide everything from the lowering logic to diagnostics, while overloading a syntax requires untangling of what's going on

Reply:
Personally, I see the `async Fn()` syntax as more confusing, because it lies about which trait is actually being implemented. It implies that it’s equivalent to `Fn() -> impl Future` (by analogy with function items), but actually `Fn()` is not implemented at all. Using a separate name helps users understand that critical distinction.

Reply:
From the user perspective there is no lie. The `AsyncFn` trait is a compiler internal implementation detail. The user wants a `Fn()` that is `async`. Just like with `fn` items, you just add `async` in front of it and get what you asked for.

I do think it was a mistake to include this compiler internal implementation detail in the RFC considering the amount of scrutiny it has gotten. It's a detail interesting to compiler developers, but not something that is actually useful to know about as a user.

Reply:
> The user wants a `Fn()` that is `async`.

That may be what the naïve, uninformed user thinks they want, but what they actually get is a `LendingFn()` that is `async`. The distinction between these concepts is the entire point of this RFC, we shouldn’t be obfuscating it.

Reply:
I don't appreciate the phrasing and don't think continueing here is productive.

Reply:
The distinction between `Fn()` and `LendingFn()` is observable, with the former you can drop the closure before its return value but that’s not possible with the latter. The APIs are distinct, and the naming should reflect that.

Reply:
> First of all, I think I've already previously said that suggesting that anyone thinks users are [stupid](https://github.com/rust-lang/rfcs/pull/3668#issuecomment-2201815984), naïve, or uninformed

All users start out as uninformed, being uninformed is the natural state of things, that's why we have intuitive APIs and good documentation.

Reply:
Isn't an intuitive API exactly what a new user expects to just work?

Or the other way around: whatever a new user tries because they've seen it else where, would be the intuitive design, as it doesn't require learning something new.


|         | function | trait   |
| ------- | -------- | ------- |
| regular | `fn`     | `Fn`    |
| async   | `async fn` | `async Fn` |

Reply:
The observable differences here are, first and foremost, due to the `async`-ness. This is a natural parallel to how `async fn` captures its arguments: `async Fn` necessarily the same thing, regardless of how it is desugared.

The specific design of the internal, unstable traits used to record this fact inside the compiler are not the important thing for the user: they already have the information they need from the `async` modifier, and they don't really benefit from getting that information in a lower-level way.

Reply:
@oli-obk: Extending the symmetry you propose one column further:

|         | function | trait   | closure |
| ------- | -------- | ------- | ------- |
| regular | `fn`     | `Fn`    | `\|\| ..` |
| async   | `async fn` | `async Fn` | `async \|\| ..` |

Reply:
Or:

```rust
fn f() -> impl async Fn() -> u8 {
    async fn g() -> u8 { 42 }
    async || g().await
}
```

Reply:
> Isn't an intuitive API exactly what a new user expects to just work?
> 
> Or the other way around: whatever a new user tries because they've seen it else where, would be the intuitive design, as it doesn't require learning something new.
> |         | function | trait   |
> | ------- | -------- | ------- |
> | regular | `fn`     | `Fn`    |
> | async   | `async fn` | `async Fn` |

One difference is that an `async fn` is still a `fn` (that returns a future), while an `async Fn` is not a `Fn`.

Reply:
> One difference is that an async fn is still a fn (that returns a future)

I feel like that difference is not as important in a larger context though? Sure: I can agree that `async fn` *is* an `fn`. But you don't just call an `fn` by itself and drop the value on the floor, and you can't just plug an `async fn foo() -> i32` into a hole that expects an `fn() -> i32`.

It's kind of a half-parallel, but I'm not necessarily convinced it's a parallel that users are really going to appreciate (or miss when it's gone).

Similarly, we probably don't need to worry about users who have an `async Fn() -> i32` and try to plug it into some `Fn() -> T`, expecting that `T = impl Future<Output = i32>`, since that's already not really going to get much mileage given the fact that the future necessarily needs to be awaited. It's similar to above.

I think the point that several others are trying to make above is users *do* know that the `async` keyword transforms the return value of a callable (fn item/closure/parameters with an `Fn` bound) in parallel ways, and that parallel is really valuable in contrast.

Reply:
> Similarly, we probably don't need to worry about users who have an `async Fn() -> i32` and try to plug it into some `Fn() -> T`, expecting that `T = impl Future<Output = i32>`

This is a totally reasonable thing to (try to) do? For example, you have a list of values, you try to `map` them to a `Future`, collect into some buffer, then join/spawn them.

Reply:
Well that's necessarily not going to work given that `async Fn` is lending, and that doesn't change depending on if `Fn` is named `AsyncFn` or `async Fn`.

Reply:
The point I'm trying to make above in [my comment](https://github.com/rust-lang/rfcs/pull/3668#discussion_r1687353322) is less about the particular ways that "lending" is observable, and more about the fact that "`async fn` items are `fn` items" overlooks the fact that one of the most important parts of the signature, the return type, is not preserved, and so suggesting that that's something that's odd about `async Fn` <=> `Fn` bounds is kinda artificial, because users can't really appreciate it without having to understand that `Future`s pop out necessarily as part of the async desugaring.

In general, I expect idiomatic async + closures APIs to use `async` closures and `async Fn` trait bounds together. So hat I *do* expect users to be able to appreciate is `async` as a modifier on all of the callable-ish things. Making sure that users only need to shift from the sync world to the `async` world by remembering to pepper their code with `async` and `await` is legitimately useful from a pedagogical standpoint.

Reply:
> Well that's necessarily not going to work

Exactly—the symmetry of `Fn`/`async Fn` with `fn`/`async fn` suggests that it should work, and yet it doesn’t. So the symmetry is actually counterproductive in this instance, misleading users

Reply:
I think that puts an incredibly large emphasis on a very specific aspect of this symmetry, so I guess we will literally just have to disagree 👍

Reply:
This "asymmetry" is also not new, or even unique to the `Fn` traits. Rather, it is the usual meaning of the `async` keyword, applied to the `Fn` traits' methods, with their `self` parameters. Thinking of it this way preserves the "`fn` that returns a future" intuition, just phrased in a way that captures the existing subtleties around lifetime capture.

Expecting `async Fn() -> i32` to satisfy `Fn() -> impl Future` is thus an *incomplete reading* of the `async fn() -> i32` -> `fn() -> impl Future` transformation. Users with this expectation have not been misled by a false symmetry- merely misunderstood the full implications of the `async` keyword.

We're also going to keep hitting this for *any* `async`-ified trait. It won't ever make sense for any `async`-ified trait to satisfy its non-`async` variant, and indeed many of them won't even have a type like `FnOnce::Output` at all. We simply can't do the proper transformation on the return type alone. (This is similar to how we can sometimes, but not always, avoid a special `try` variant of some APIs.)